### ISSUE TITLE
Generate page from json directly

### ISSUE THREAD

**Issue by kuzvac** [MoOx/phenomic#358]
Can i use it for display pages from json directly, not from markdown?
Use case: I get json that need statinamic from wordpress (through wp-api, and some modifications). Feed this json to statinamic (route setups manually or dynamic), and statinamic render my pages.
It's really to do?

**Comment by thangngoc89** [comment]
Well. Thanks for a interesting feature request. This is currently not possible with Statinamic. But there is workaround, you can save your json data to files and let's Statinamic handle the rest like normal.

**Comment by kuzvac** [comment]
@thangngoc89 hmm, interesting, thanks for suggesting. Save json data to json file or create markdown from json?
I think i dive into it at end of next week.

**Comment by thangngoc89** [comment]
> Save json data to json file or create markdown from json

That's depends on which kind of data. Could you show me an example data structure of the JSON ?

**Comment by kuzvac** [comment]
Real json date i can provide in next week. For now i have schema of fields which wp-api can provide: 
post - http://v2.wp-api.org/reference/posts/
pages - http://v2.wp-api.org/reference/pages/

**Comment by thangngoc89** [comment]
> For now i have schema of fields

That's more than enough. You have 2 options here: 

- Easy: Dump JSON data into files like this:

```md
---json
// Dump all data you need here with JSON.stringify()
---

Main content put in there
```

- Write a content-loader that doesn't parse frontmatter but read and emit files directly from those JSON field. This is hard if you don't know about how Statinamic works internally but it might saves you a little performances if you are working with a large amount of files.

**Comment by MoOx** [comment]
If you want to have statinamic working from with static/client build, you can simply generate files from your wp-api with the following content

```
---
# header (json/yml/whatever) with title, route, date etc
---

html content (since wordpress wysiwyg generate html)
```

This way, you can consume this files via content-loader with an empty rendering method (since body will already be html). statinamic will be able to generate collection data from those files.
Seems really simple to do, and I am interested since I maintain a wordpress website (admin is easier for my wife than raw markdown files).

For assets (wordpress media), it should not be an issue since wordpress generate full url for media in the wysiwyg.

A script that generate from wp-api to statinamic ready files will be highly appreciated for me (and the community haha).

**Comment by thangngoc89** [comment]
For the record, I'm building something similar to you. Take posts from Discourse forum API and build static site with that. If you need a hint, take a look 

https://github.com/thangngoc89/dnh-blog

(most interested parts is in `/data/`)

**Comment by MoOx** [comment]
@kuzvac any update maybe?

**Comment by kuzvac** [comment]
No, for now, we built layout pages with static data. 
But we will be doing it later. I write here about progress.

**Comment by MoOx** [comment]
Now we are supposed to be able to do this via phenomic loader plugins. Will need some tests I guess. https://phenomic.io/docs/usage/plugins/

**Comment by facelesspanda** [comment]
@MoOx I was looking for a way to handle data files and render them with react components, and looked at Phenomic's plugin pipeline, but the PhenomicLoader only deals with front-matter and markdown, in your docs, and it's documented that "body is what will be injected in Phenomic's `<BodyContainer>`" It's okay by me, but 1) yaml and json don't have front-matter, they directly have matter, and have no body, and 2) I don't want to output html into a body in my plugin, I want to pass data to my React components so that they render it as an html DOM.

**Comment by MoOx** [comment]
We can totally adjust our loader to accept file like yml and json directly.
I am also thinking about removing the head/body thing, and just directly have the data, probably keeping a "body" field generated from text file using a front-matter + some content.
The BodyContainer only make sense to display a body indeed, but you can skip it's usage and only rely on data to display whatever you want. A workaround I am currently using is to use .md files with just front-matter (eg: for showcases on our website)

**Comment by facelesspanda** [comment]
Yeah, as a work-around, it's a good one, since if you ever want to add a markdown content, you easily can without changing your structure that much.
I'm actively looking for a way to build a static site like this:
```
front-matter===|
               |=> js objects==|
markdown=======|               |
                               |=> site data structure ===|
                   yaml/json===|                          |
                                                  ________V_________
                   Routing parameters ===========>|React components|
                                                  ==================
                                                          |
                                                          |
                                                          V
                                                  ===================
                                                  |Static HTML files|
                                                  ===================
```

Ideally, with means to smartly parse the source file paths into data (something like `/animals/birds/sparrow.yml` would become, after loading, a object in the Content Tree with a `category: bird` property, viewable at url `/animals/sparrow`, because its Content Type defined something in the likes
```yaml
route: {
    from: 'animals/:category/:name',
    to: 'animals/:name',
    save: [
        'category',
        'name'
    ]
}
```


Phenomic is the existing technology closest to allowing me to do that, yet I don't really know if I can leverage Phenomic to do it, or if I'll end up having to write my own webpack loaders and plugins anyway.

My guts say I could, but my brains, rummaging through the docs and the code, can't find a solution :/

**Comment by MoOx** [comment]
To make it easy we can add native support for raw data. Anyway, the current way to consume data & collection will be refactored and improved, and API will probably change a little bit.
For the transformation you are talking (category from file path to metadata), you can probably write a custom loader plugin.
For the routing, when data will be in the phenomic global collection, it should not be a problem since we have what I think is a nice way to create dynamic routes https://phenomic.io/docs/usage/routing/

**Comment by facelesspanda** [comment]
I guess you're right, Phenomic gives me all the tools I need, I just need to engineer my way through the problem.

Curiosity satisfied regarding raw data transformation. Thanks!

**Comment by MoOx** [comment]
If you want to take a look and play with things a little bit, here is where current transformations are made https://github.com/MoOx/phenomic/blob/38662d0561fb9fb92d074c958800433d32e3e7b9/src/loader/index.js#L31-L38. Should not be very hard to adjust this to skip frontMatter in certain situation (file extension?)

**Comment by bloodyowl** [comment]
Going to be fixed with #925
